spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer.java

```diff
@@ -1500,13 +1536,21 @@ private void cleanUpAfterInvoke(@Nullable RabbitResourceHolder resourceHolder, C
 	 * Exception thrown from listener will be wrapped to {@link ListenerExecutionFailedException}.
 	 *
 	 * @param listener the Rabbit MessageListener to invoke
-	 * @param message the received Rabbit Message
+	 * @param data the received Rabbit Message or List of Message.
 	 *
 	 * @see org.springframework.amqp.core.MessageListener#onMessage
 	 */
-	protected void doInvokeListener(MessageListener listener, Message message) {
+	@SuppressWarnings("unchecked")
+	protected void doInvokeListener(MessageListener listener, Object data) {
+		Message message = null;
 		try {
-			listener.onMessage(message);
+			if (listener instanceof BatchMessageListener) {
```

Comment:
Don't we need to do something slightly opposite?
Check for `data instance of `List` and always call `onMessageBatch()` since it is there on the `MessageListener` anyway.